src/librustc_mir/transform/inline.rs

```diff
@@ -0,0 +1,730 @@
+                        if let ty::TyFnDef(callee_def_id, substs, _) = f.ty.sty {
+                            // Don't inline the same function multiple times.
+                            if callsite.callee != callee_def_id {
+                                callsites.push(CallSite {
+                                    caller: callsite.caller,
+                                    callee: callee_def_id,
+                                    substs: substs,
+                                    bb: bb,
+                                    location: terminator.source_info
+                                });
+                            }
+                        }
+                    }
+                }
+
+
+                csi -= 1;
+                if scc.len() == 1 {
+                    callsites.swap_remove(csi);
+                } else {
+                    callsites.remove(csi);
+                }
+
+                local_change = true;
+                changed = true;
+            }
+
+            if !local_change {
+                break;
+            }
+        }
+
+        changed
+    }
+
+    fn get_foreign_mir(&mut self, def_id: DefId) -> Option<Rc<Mir<'tcx>>> {
+        if let Some(mir) = self.foreign_mirs.get(&def_id).cloned() {
+            return Some(mir);
+        }
+        // Cache the foreign MIR
+        let mir = self.tcx.sess.cstore.maybe_get_item_mir(self.tcx, def_id);
+        let mir = mir.map(Rc::new);
+        if let Some(ref mir) = mir {
+            self.foreign_mirs.insert(def_id, mir.clone());
+        }
+
+        mir
+    }
+
+    fn should_inline(&self, callsite: CallSite<'tcx>,
+                     callee_mir: &'a Mir<'tcx>) -> bool {
+
+        let tcx = self.tcx;
+
+        // Don't inline closures
+        if callee_mir.upvar_decls.len() > 0 {
+            return false;
+        }
+
+        // Don't inline calls to trait methods
+        // FIXME: Should try to resolve it to a concrete method, and
+        // only bail if that isn't possible
+        let trait_def = tcx.trait_of_item(callsite.callee);
+        if trait_def.is_some() { return false; }
+
+        let attrs = tcx.get_attrs(callsite.callee);
+        let hint = attr::find_inline_attr(None, &attrs[..]);
+
+        let hinted = match hint {
+            // Just treat inline(always) as a hint for now,
+            // there are cases that prevent unwinding that we
+            // need to check for first.
+            attr::InlineAttr::Always => true,
+            attr::InlineAttr::Never => return false,
+            attr::InlineAttr::Hint => true,
+            attr::InlineAttr::None => false,
+        };
+
+        // Only inline local functions if they would be eligible for
+        // cross-crate inlining. This ensures that any symbols they
+        // use are reachable cross-crate
+        // FIXME: This shouldn't be necessary, trans should generate
+        // the reachable set from the MIR.
+        if callsite.callee.is_local() {
```

Comment:
This is more conservative than necessary, right? If the current function isn't eligible for cross-crate inlining, then it shouldn't matter whether the callee is eligible for cross-crate inlining.

If so, you don't need to fix this now—in fact, it may be better if you *don't* fix it, to make this patch more minimal. But it would be nice to add a `FIXME` here.